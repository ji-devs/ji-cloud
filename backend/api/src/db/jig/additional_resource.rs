use shared::domain::jig::{AdditionalResourceId, JigId};
use sqlx::PgPool;

pub async fn create(
    pool: &PgPool,
    parent: JigId,
    url: String,
) -> anyhow::Result<AdditionalResourceId> {
    sqlx::query!(
        r#"
insert into jig_additional_resource (jig_id, url)
values ($1, $2)
returning id
        "#,
        parent.0,
        url,
    )
    .fetch_one(pool)
    .await
    .map(|it| AdditionalResourceId(it.id))
    .map_err(Into::into)
}

pub async fn get(
    pool: &PgPool,
    parent: JigId,
    id: AdditionalResourceId,
) -> anyhow::Result<Option<String>> {
    let url = sqlx::query!(
        r#"
select
    url as "url!: String"
from jig_additional_resource
where jig_id = $1 and id = $2
        "#,
        parent.0,
        id.0,
    )
    .fetch_optional(pool)
    .await?
    .map(|it| it.url);

    Ok(url)
}

pub async fn update(
    pool: &PgPool,
    parent_id: JigId,
    id: AdditionalResourceId,
    url: Option<String>,
) -> anyhow::Result<bool> {
    let mut txn = pool.begin().await?;

    sqlx::query!(
        r#"
update jig_additional_resource
set url = coalesce($3, url)
where jig_id = $1 and id = $2
        "#,
        parent_id.0,
        id.0,
        url,
    )
    .execute(&mut txn)
    .await?;

    txn.commit().await?;

    Ok(true)
}

pub async fn delete(
    pool: &PgPool,
    parent_id: JigId,
    id: AdditionalResourceId,
) -> anyhow::Result<()> {
    let mut txn = pool.begin().await?;

    sqlx::query!(
        r#"
delete from jig_additional_resource
where jig_id = $1 and id = $2
        "#,
        parent_id.0,
        id.0,
    )
    .execute(&mut txn)
    .await?;

    txn.commit().await?;

    Ok(())
}
