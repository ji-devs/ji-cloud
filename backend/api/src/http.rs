use std::{net::TcpListener, sync::Arc};

use actix_service::Service;
use actix_web::{
    dev::{MessageBody, Server, ServiceRequest, ServiceResponse},
    error::Error,
    middleware::{Compat, Condition},
    web::{method, Data},
    HttpResponse,
};
use core::{
    config::JSON_BODY_LIMIT,
    env::env_bool,
    http::{get_addr, get_tcp_fd},
    settings::RuntimeSettings,
};
use futures::Future;
use sqlx::postgres::PgPool;
use tracing::Span;
use tracing_actix_web::{root_span, DefaultRootSpanBuilder, RootSpanBuilder, TracingLogger};

use crate::{
    error::BasicError,
    service::{self, mail, s3, upload::cleaner, ServiceData},
    translate,
};

mod cors;
mod endpoints;

fn log_ise<B: MessageBody, T>(
    request: ServiceRequest,
    srv: &T,
) -> impl Future<Output = actix_web::Result<T::Response>>
where
    T: Service<ServiceRequest, Response = ServiceResponse<B>, Error = actix_web::Error>,
{
    let uri: serde_json::Value = request.uri().to_string().into();
    let method: serde_json::Value = request.method().to_string().into();

    let future = srv.call(request);
    async {
        let mut result = future.await?;
        if result.status() == 500 {
            let response: &mut actix_web::HttpResponse<_> = result.response_mut();

            if let Some(err) = response.extensions_mut().remove::<anyhow::Error>() {
                log::error!("ISE while responding to request: {:?}", err);
                sentry::add_breadcrumb(sentry::Breadcrumb {
                    ty: "http".to_owned(),
                    category: Some("request".into()),
                    data: {
                        let mut map = sentry::protocol::Map::new();
                        map.insert("url".to_owned(), uri);
                        map.insert("method".to_owned(), method);
                        map
                    },
                    ..Default::default()
                });

                sentry::integrations::anyhow::capture_anyhow(&err);
            }
        }

        Ok(result)
    }
}

pub struct Application {
    port: u16,
    server: Option<Server>,
}

impl Application {
    pub fn new(port: u16, server: Server) -> Self {
        Self {
            port,
            server: Some(server),
        }
    }

    pub fn port(&self) -> u16 {
        self.port
    }

    // A more expressive name that makes it clear that
    // this function only returns when the application is stopped.
    pub async fn run_until_stopped(mut self) -> Result<(), std::io::Error> {
        if let Some(server) = self.server.take() {
            server.await?;
        }

        Ok(())
    }

    pub async fn stop(mut self, graceful: bool) {
        if let Some(server) = self.server.take() {
            server.stop(graceful).await
        }
    }
}

impl Drop for Application {
    fn drop(&mut self) {
        if let Some(server) = self.server.take() {
            let _ = tokio::spawn(server.stop(false));
        }
    }
}

#[actix_web::main]
pub async fn build_and_run(
    pool: PgPool,
    settings: RuntimeSettings,
    s3: Option<s3::Client>,
    gcp_key_store: Option<service::GcpAccessKeyStore>,
    gcs: Option<service::storage::Client>,
    algolia: Option<crate::algolia::Client>,
    algolia_key_store: Option<crate::algolia::SearchKeyStore>,
    jwk_verifier: Arc<crate::jwk::JwkVerifier>,
    mail_client: Option<mail::Client>,
    algolia_manager: Option<crate::algolia::Manager>,
    media_upload_cleaner: Option<cleaner::UploadCleaner>,
    google_translate: Option<translate::GoogleTranslate>,
) -> anyhow::Result<()> {
    let app = build(
        pool,
        settings,
        s3,
        gcp_key_store,
        gcs,
        algolia,
        algolia_key_store,
        jwk_verifier,
        mail_client,
        algolia_manager,
        media_upload_cleaner,
        google_translate,
    )?;
    app.run_until_stopped().await?;

    Ok(())
}

pub fn build(
    pool: PgPool,
    settings: RuntimeSettings,
    s3: Option<s3::Client>,
    gcp_key_store: Option<service::GcpAccessKeyStore>,
    gcs: Option<service::storage::Client>,
    algolia: Option<crate::algolia::Client>,
    algolia_key_store: Option<crate::algolia::SearchKeyStore>,
    jwk_verifier: Arc<crate::jwk::JwkVerifier>,
    mail_client: Option<mail::Client>,
    algolia_manager: Option<crate::algolia::Manager>,
    media_upload_cleaner: Option<cleaner::UploadCleaner>,
    google_translate: Option<translate::GoogleTranslate>,
) -> anyhow::Result<Application> {
    let local_insecure = settings.is_local();
    let api_port = settings.api_port;

    let s3 = s3.map(ServiceData::new);
    let gcp_key_store = gcp_key_store.map(ServiceData::new);
    let gcs = gcs.map(ServiceData::new);
    let algolia = algolia.map(ServiceData::new);
    let algolia_key_store = algolia_key_store.map(ServiceData::new);
    let mail_client = mail_client.map(ServiceData::new);
    let algolia_manager = algolia_manager.map(ServiceData::new);
    let media_upload_cleaner = media_upload_cleaner.map(ServiceData::new);
    let google_translate = google_translate.map(ServiceData::new);

    let server = actix_web::HttpServer::new(move || {
        let app = actix_web::App::new()
            .app_data(Data::new(pool.clone()))
            .app_data(Data::new(settings.clone()));

        let app = match s3.clone() {
            Some(s3) => app.app_data(s3),
            None => app,
        };

        let app = match gcp_key_store.clone() {
            Some(gcp_key_store) => app.app_data(gcp_key_store),
            None => app,
        };

        let app = match gcs.clone() {
            Some(gcs) => app.app_data(gcs),
            None => app,
        };

        let app = match algolia.clone() {
            Some(algolia) => app.app_data(algolia),
            None => app,
        };

        let app = match algolia_key_store.clone() {
            Some(algolia_key_store) => app.app_data(algolia_key_store),
            None => app,
        };

        let app = match mail_client.clone() {
            Some(mail_client) => app.app_data(mail_client),
            None => app,
        };

        let app = match algolia_manager.clone() {
            Some(algolia_manager) => app.app_data(algolia_manager),
            None => app,
        };

        let app = match media_upload_cleaner.clone() {
            Some(media_upload_cleaner) => app.app_data(media_upload_cleaner),
            None => app,
        };
        let app = match google_translate.clone() {
            Some(google_translate) => app.app_data(google_translate),
            None => app,
        };

        let enable_tracing_logs = env_bool("ENABLE_TRACING_LOGS");

        app.app_data(Data::from(jwk_verifier.clone()))
            .wrap(cors::get(local_insecure))
            .wrap(Condition::new(
                !enable_tracing_logs,
                Compat::new(actix_web::middleware::Logger::default()),
            ))
            .wrap(Condition::new(
                enable_tracing_logs,
                Compat::new(TracingLogger::<JigziSpanBuilder>::new()),
            ))
            .wrap_fn(log_ise)
            .app_data(
                actix_web::web::JsonConfig::default()
                    .limit(JSON_BODY_LIMIT as usize)
                    .error_handler(|_, _| bad_request_handler()),
            )
            .app_data(
                actix_web::web::QueryConfig::default().error_handler(|_, _| bad_request_handler()),
            )
            .app_data(
                actix_web::web::PathConfig::default().error_handler(|_, _| bad_request_handler()),
            )
            .external_resource(
                "google_cloud_oauth",
                "https://accounts.google.com/o/oauth2/v2/auth",
            )
            .default_service(actix_web::web::to(default_route))
            .configure(endpoints::user::configure)
            .configure(endpoints::category::configure)
            .configure(endpoints::image::configure)
            .configure(endpoints::audio::configure)
            .configure(endpoints::meta::configure)
            .configure(endpoints::jig::configure)
            .configure(endpoints::module::configure)
            .configure(endpoints::additional_resource::configure)
            .configure(endpoints::jig::report::configure)
            .configure(endpoints::jig::curation::configure)
            .configure(endpoints::course::configure)
            .configure(endpoints::admin::configure)
            .configure(endpoints::animation::configure)
            .configure(endpoints::search::configure)
            .configure(endpoints::media::configure)
            .configure(endpoints::session::configure)
            .configure(endpoints::locale::configure)
            .configure(endpoints::scheduler::configure)
            .configure(endpoints::pdf::configure)
            .configure(endpoints::badge::configure)
            .route("/", method(http::Method::GET).to(no_content_response))
    });

    // if listenfd doesn't take a TcpListener (i.e. we're not running via
    // the command above), we fall back to explicitly binding to a given
    // host:port.
    let listener = if let Some(l) = get_tcp_fd() {
        l
    } else {
        TcpListener::bind(get_addr(Some(api_port)))?
    };

    let port = listener.local_addr().unwrap().port();

    let server = server.listen(listener)?;

    Ok(Application::new(port, server.run()))
}

fn default_route() -> HttpResponse {
    HttpResponse::NotFound().json(BasicError::with_message(
        http::StatusCode::NOT_FOUND,
        "Route not found".to_owned(),
    ))
}

async fn no_content_response() -> HttpResponse {
    HttpResponse::NoContent().finish()
}

pub fn bad_request_handler() -> actix_web::Error {
    BasicError::new(http::StatusCode::BAD_REQUEST).into()
}

pub struct JigziSpanBuilder;

impl RootSpanBuilder for JigziSpanBuilder {
    fn on_request_start(request: &ServiceRequest) -> Span {
        let route = request.match_pattern().unwrap_or("default".to_string());
        let transaction_name = format!("{} {}", request.method(), route);
        let transaction_name = transaction_name.as_str();
        root_span!(request, message = transaction_name)
    }

    fn on_request_end<B>(span: Span, outcome: &Result<ServiceResponse<B>, Error>) {
        DefaultRootSpanBuilder::on_request_end(span, outcome)
    }
}
