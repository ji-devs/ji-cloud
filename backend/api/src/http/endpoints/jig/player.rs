use actix_web::{
    web::{self, Data, Json},
    HttpResponse,
};
use shared::{
    api::{endpoints::jig::player, ApiEndpoint},
    domain::jig::{
        player::{JigPlayerSession, JigPlayerSessionListResponse},
        JigId,
    },
};
use sqlx::PgPool;

use crate::{db, error, extractor::TokenUser};

/// Create a jig player session for the author, if one does not exist already.
pub async fn create(
    db: Data<PgPool>,
    claims: TokenUser,
    req: Json<<player::Create as ApiEndpoint>::Req>,
) -> Result<HttpResponse, error::JigCode> {
    let req = req.into_inner();

    // let settings = req.settings.clone(); // FIXME get rid of this clone

    db::jig::authz(&*db, claims.0.user_id, Some(req.jig_id.clone())).await?;

    let index = db::jig::player::create(&db, req.jig_id, &req.settings).await?;

    Ok(HttpResponse::Created().json(JigPlayerSession {
        index,
        settings: req.settings,
    }))
}

/// Fetch a jig player session code from it's jig if it exists.
pub async fn list(
    db: Data<PgPool>,
    _claims: TokenUser,
    path: web::Path<JigId>,
) -> Result<Json<<player::List as ApiEndpoint>::Res>, error::JigCode> {
    let id = path.into_inner();

    let sessions = db::jig::player::list_sessions(&*db, id).await?;

    Ok(Json(JigPlayerSessionListResponse { sessions }))
}

pub mod instance {
    use actix_web::{
        web::{Data, Json},
        HttpResponse,
    };
    use chrono::{Duration, Utc};
    use core::settings::RuntimeSettings;
    use serde::Deserialize;
    use shared::{
        api::{endpoints::jig::player, ApiEndpoint},
        domain::jig::player::instance::PlayerSessionInstanceResponse,
    };
    use sqlx::PgPool;

    use crate::{
        db, error,
        extractor::{IPAddress, UserAgent},
        token::{create_player_session_instance_token, validate_token},
    };
    use uuid::Uuid;

    /// Create a jig player session instance
    pub async fn create_session_instance(
        settings: Data<RuntimeSettings>,
        db: Data<PgPool>,
        ip_address: IPAddress,
        user_agent: UserAgent,
        req: Json<<player::instance::Create as ApiEndpoint>::Req>,
    ) -> Result<
        (
            Json<<player::instance::Create as ApiEndpoint>::Res>,
            actix_web::http::StatusCode,
        ),
        error::JigCode,
    > {
        let req = req.into_inner();

        let resp =
            db::jig::player::create_session_instance(&*db, req.index, ip_address, user_agent)
                .await?;

        let token: String = create_player_session_instance_token(
            &settings.token_secret,
            Duration::weeks(2),
            &resp.2,
            Utc::now(),
        )?;

        Ok((
            Json(PlayerSessionInstanceResponse {
                jig_id: resp.0,
                settings: resp.1,
                token,
            }),
            actix_web::http::StatusCode::CREATED,
        ))
    }

    #[derive(Deserialize)]
    struct InstanceToken {
        /// The instance this token is for.
        pub sub: Uuid,
    }

    /// Create a jig player session for someone who's not the author, if one doesn't already exist
    pub async fn complete_session_instance(
        settings: Data<RuntimeSettings>,
        db: Data<PgPool>,
        ip_address: IPAddress,
        user_agent: UserAgent,
        req: Json<<player::instance::Complete as ApiEndpoint>::Req>,
    ) -> Result<HttpResponse, error::JigCode> {
        let req = req.into_inner();

        let token = validate_token(&req.token, None, &settings.token_secret)
            .map_err(|_| error::JigCode::Forbidden)?;

        let instance_token: InstanceToken = serde_json::from_value(token)?;

        db::jig::player::complete_session_instance(
            &db,
            req.jig_id,
            ip_address,
            user_agent,
            instance_token.sub,
        )
        .await?;

        Ok(HttpResponse::NoContent().finish())
    }
}
