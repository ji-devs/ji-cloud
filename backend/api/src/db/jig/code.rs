use crate::error;
use config::JIG_CODE_NUMBER_OF_DIGITS;
use shared::domain::jig::JigId;
use sqlx::postgres::PgDatabaseError;
use sqlx::PgPool;
use uuid::Uuid;

pub async fn create(db: &PgPool, jig_id: JigId) -> Result<i16, error::JigCode> {
    let mut txn = db.begin().await?;

    let exists = sqlx::query!(
        r#"select exists(select 1 from jig_code where jig_id=$1) as "exists!""#,
        jig_id.0
    )
    .fetch_one(&mut txn)
    .await?
    .exists;
    if exists {
        txn.rollback().await?;
        return Err(error::JigCode::Conflict);
    }

    // retry until successful:
    let mut index = hash_id_to_code(jig_id.0);
    let mut n: i16 = 0;

    loop {
        match sqlx::query!(
            //language=SQL
            r#"insert into jig_code (jig_id, index) values ($1, $2)"#,
            jig_id.0,
            index
        )
        .execute(&mut txn)
        .await
        {
            Ok(_) => return Ok(index), // insert successful
            Err(err) => match err {
                sqlx::Error::Database(db_err) => {
                    let constraint = db_err.downcast_ref::<PgDatabaseError>().constraint();
                    match constraint {
                        Some("jig_code_jig_id_key") => {
                            // code exists for the jig!
                            txn.rollback().await?;
                            return Err(error::JigCode::Conflict);
                        }
                        Some("jig_code_index_key") => {
                            // same code but different id. retry insert with a new code
                            index = rehash(index, n);
                            n += 1;
                        }
                        Some("jig_code_jig_id_fkey") => {
                            txn.rollback().await?;
                            return Err(error::JigCode::ResourceNotFound);
                        }
                        db_err => {
                            txn.rollback().await?;
                            return Err(anyhow::anyhow!(
                                "{}",
                                db_err.unwrap_or("unknown database error")
                            )
                            .into());
                        }
                    }
                }
                err => return Err(err.into()),
            },
        }
    }
}

/// Hashes a Uuid by
/// 1. XORing every 2 bytes together as an i16,
/// 2. clamping to within the digit requirement using mod (4 digits here),
/// 3. taking the absolute value to get rid of negative numbers
fn hash_id_to_code(id: Uuid) -> i16 {
    let hash = id
        .as_bytes()
        .windows(2)
        .fold(0, |acc, w| acc ^ (w[0] as i16 + (w[1] as i16) << 8));

    let result = (hash % 10_i16.pow(JIG_CODE_NUMBER_OF_DIGITS)) as i16;

    result.abs()
}

/// Rehash by
/// 1. adding 2.pow(attempt_number) to the previous conflicting hash:
///      hash(j) = hash(j-1) + 2.pow(j-1),
/// 2. clamping to within the digit requirement using mod (4 digits here),
/// 3. taking the absolute value to get rid of negative numbers
fn rehash(hash: i16, attempt: i16) -> i16 {
    (hash + (1 << attempt)) % 10_i16.pow(JIG_CODE_NUMBER_OF_DIGITS).abs()
}

pub async fn get(db: &PgPool, jig_id: JigId) -> anyhow::Result<Option<i16>> {
    let index = sqlx::query!(
        r#"select index as "index!: i16" from jig_code where jig_id = $1"#,
        jig_id.0
    )
    .fetch_optional(db)
    .await?
    .map(|it| it.index);

    Ok(index)
}

pub async fn get_jig_from_code(db: &PgPool, index: i16) -> anyhow::Result<Option<JigId>> {
    let id: Option<JigId> = sqlx::query!(
        r#"select jig_id as "id!: JigId" from jig_code where index = $1"#,
        index
    )
    .fetch_optional(db)
    .await?
    .map(|it| it.id);

    log::info!("{}", id.unwrap().0);

    Ok(id)
}
