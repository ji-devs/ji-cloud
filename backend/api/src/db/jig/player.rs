use crate::error;
use crate::extractor::IPAddress;
use core::config::JIG_PLAYER_SESSION_CODE_MAX;
use shared::domain::jig::player::JigPlayerSettings;
use shared::domain::jig::{JigId, TextDirection};
use sqlx::postgres::PgDatabaseError;
use sqlx::PgPool;
use uuid::Uuid;

pub async fn create(
    db: &PgPool,
    jig_id: JigId,
    settings: JigPlayerSettings,
) -> Result<i16, error::JigCode> {
    let mut txn = db.begin().await?;

    let exists = sqlx::query!(
        r#"select exists(select 1 from jig_player_session where jig_id=$1) as "exists!""#,
        jig_id.0
    )
    .fetch_one(&mut txn)
    .await?
    .exists;
    if exists {
        txn.rollback().await?;
        return Err(error::JigCode::Conflict);
    }

    // retry until successful:
    let mut index = hash_id_to_code(jig_id.0);
    let mut n: i16 = 0;

    loop {
        match sqlx::query!(
            //language=SQL
            r#"
insert into jig_player_session (jig_id, index, direction, display_score, track_assessments, drag_assist)
values ($1, $2, $3, $4, $5, $6)

"#,
            jig_id.0,
            index,
            settings.direction as i16,
            settings.display_score,
            settings.track_assessments,
            settings.drag_assist,
        )
        .execute(&mut txn)
        .await
        {
            Ok(_) => { // insert successful
                txn.commit().await?;
                return Ok(index);
            },
            Err(err) => match err {
                sqlx::Error::Database(db_err) => {
                    let constraint = db_err.downcast_ref::<PgDatabaseError>().constraint();
                    match constraint {
                        Some("jig_player_session_jig_id_key") => {
                            // session exists for the jig, prevents race condition?
                            log::info!("1");
                            txn.rollback().await?;
                            return Err(error::JigCode::Conflict);
                        }
                        Some("jig_player_session_pkey") => {
                            // same code but different jig. retry insert with a new code
                            log::info!("how? {}", &index);
                            index = rehash(index, &n);
                            n += 1;
                        }
                        Some("jig_player_session_jig_id_fkey") => {
                            // no jig with this id exists
                            log::info!("2");
                            txn.rollback().await?;
                            return Err(error::JigCode::ResourceNotFound);
                        }
                        db_err => {
                            txn.rollback().await?;
                            log::info!("3");
                            return Err(anyhow::anyhow!(
                                "{}",
                                db_err.unwrap_or("unknown database error")
                            )
                            .into());
                        }
                    }
                }
                err => return Err(anyhow::anyhow!("sqlx error: {:?}", err).into()),
            },
        }
    }
}

/// Creates new jig player session for a player
// TODO: Fixme
// TODO: Add Settings?
pub async fn create_user_session(
    db: &PgPool,
    session_index: i16,
    ip_addr: IPAddress,
) -> Result<String, error::JigCode> {
    let mut txn = db.begin().await?;

    let jig_id = sqlx::query!(
        //language=SQL
        r#"
        select jig_id
        from jig_player_session
        where index=$1
        "#,
        session_index
    )
    .fetch_optional(&mut txn)
    .await?
    .ok_or(error::JigCode::ResourceNotFound)?;

    // get the user agent header and ip address
    let ip_address = ip_addr.ip_addr.ok_or(error::JigCode::ResourceNotFound)?;
    let user_agent = ip_addr.user_agent.ok_or(error::JigCode::ResourceNotFound)?;

    // insert into the jig_player_session_instance table returning the instance_id
    let jig_player_session_instance_id = sqlx::query!(
        //language=SQL
        r#"
        insert into jig_player_session_instance (index, jig_id, ip_addr, user_agent)
        values ($1, $2, $3, $4)
        returning instance_id as "id: Uuid"
        "#,
        session_index,
        jig_id.jig_id,
        ip_address,
        user_agent
    )
    .fetch_one(&mut txn)
    .await?;

    txn.commit().await?;

    Ok(jig_player_session_instance_id.id.to_string())
}

/// Completes a jig player session for a player and updates play count
pub async fn complete_user_session(
    db: &PgPool,
    jig_id: JigId,
    ip_addr: IPAddress,
    session_id: &str,
) -> Result<(), error::JigCode> {
    let txn = db.begin().await?;

    let instance_id = Uuid::parse_str(&session_id).unwrap();

    let jig_session_index_query = sqlx::query!(
        //language=SQL
        r#"
        select *
        from jig_player_session_instance
        where instance_id = $1
        "#,
        instance_id,
    )
    .fetch_optional(db)
    .await?
    .ok_or(error::JigCode::ResourceNotFound)?;

    // get the user agent header and ip address and check validity
    let ip_address = ip_addr.ip_addr.ok_or(error::JigCode::ResourceNotFound)?;
    let user_agent = ip_addr.user_agent.ok_or(error::JigCode::ResourceNotFound)?;

    // unwrap because invalid ip/user agents are checked before putting them into the db
    // confirms user agent and ip address match
    if (jig_session_index_query.user_agent.unwrap()).ne(&user_agent)
        | (jig_session_index_query.ip_addr.unwrap()).ne(&ip_address)
    {
        return Err(error::JigCode::ResourceNotFound);
    }

    sqlx::query!(
        //language=SQL
        r#"
        update jig_play_count
        set play_count = play_count + 1
        where jig_id = $1
        "#,
        jig_id.0,
    )
    .execute(db)
    .await?;

    txn.commit().await?;

    Ok(())
}

/// Hashes a Uuid by
/// 1. XORing every 2 bytes together as an i16,
/// 2. clamping to within the digit requirement using mod (4 digits here),
/// 3. taking the absolute value to get rid of negative numbers
fn hash_id_to_code(id: Uuid) -> i16 {
    let bytes_to_word = |a: &u8, b: &u8| *a as i16 + (*b as i16) << 8;

    let hash = id
        .as_bytes()
        .windows(2)
        .fold(0, |acc, w| acc ^ bytes_to_word(&w[0], &w[1]));

    let result = (hash % 10000) as i16;

    result.abs()
}

/// Rehash by
/// 1. adding 2.pow(attempt_number) to the previous conflicting hash:
///      hash(j) = hash(j-1) + 2.pow(j-1),
/// 2. clamping to within the digit requirement using mod (4 digits here),
/// 3. taking the absolute value to get rid of negative numbers
#[inline]
fn rehash(hash: i16, attempt: &i16) -> i16 {
    ((hash + (1 << attempt)) % (JIG_PLAYER_SESSION_CODE_MAX + 1)).abs()
}

pub async fn get(db: &PgPool, index: i16) -> anyhow::Result<Option<(JigId, JigPlayerSettings)>> {
    let res = sqlx::query!(
        //language=SQL
        r#"
select jig_id as "id: JigId",
       direction as "direction: TextDirection",
       display_score, 
       track_assessments, 
       drag_assist
from jig_player_session 
where index = $1
        "#,
        index
    )
    .fetch_optional(db)
    .await?
    .map(|it| {
        (
            it.id,
            JigPlayerSettings {
                direction: it.direction,
                display_score: it.display_score,
                track_assessments: it.track_assessments,
                drag_assist: it.drag_assist,
            },
        )
    });

    Ok(res)
}

pub async fn get_code(db: &PgPool, jig_id: JigId) -> anyhow::Result<Option<i16>> {
    log::info!("asdasda {:?}", &jig_id);
    let index = sqlx::query!(
        r#"select index as "index!: i16" from jig_player_session where jig_id = $1"#,
        jig_id.0
    )
    .fetch_optional(db)
    .await?
    .map(|it| it.index);

    log::info!("{:?}", index);

    Ok(index)
}
