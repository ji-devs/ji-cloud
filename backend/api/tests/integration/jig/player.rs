use crate::{
    fixture::Fixture,
    helpers::{initialize_server, LoginExt},
};
use http::StatusCode;
use shared::domain::jig::player::{
    instance::PlayerSessionInstanceResponse, JigPlayerSession, JigPlayerSessionListResponse,
};
use sqlx::postgres::{PgConnectOptions, PgPoolOptions};

#[sqlx::test]
async fn list(pool_opts: PgPoolOptions, conn_opts: PgConnectOptions) -> anyhow::Result<()> {
    let app = initialize_server(
        &[Fixture::MetaKinds, Fixture::User, Fixture::Jig],
        &[],
        pool_opts,
        conn_opts,
    )
    .await;

    let port = app.port();

    tokio::spawn(app.run_until_stopped());

    let client = reqwest::Client::new();

    let resp = client
        .get(&format!(
            "http://0.0.0.0:{}/v1/jig/0cc084bc-7c83-11eb-9f77-e3218dffb008/player",
            port
        ))
        .login()
        .send()
        .await?;

    assert_eq!(resp.status(), StatusCode::OK);

    let body: JigPlayerSessionListResponse = resp.json().await?;

    insta::assert_json_snapshot!(body, { ".**.expires_at" => "[timestamp]" });

    Ok(())
}

#[sqlx::test]
async fn create(pool_opts: PgPoolOptions, conn_opts: PgConnectOptions) -> anyhow::Result<()> {
    let app = initialize_server(
        &[Fixture::MetaKinds, Fixture::User, Fixture::Jig],
        &[],
        pool_opts,
        conn_opts,
    )
    .await;

    let port = app.port();

    tokio::spawn(app.run_until_stopped());

    let client = reqwest::Client::new();

    let resp = client
        .post(&format!("http://0.0.0.0:{}/v1/jig/player", port))
        .json(&serde_json::json!({
            "jigId": "3a71522a-cd77-11eb-8dc1-af3e35f7c743",
            "settings": {
                "direction": "rtl",
                "display_score": false,
                "track_assessments": false,
                "drag_assist": false,
            }
        }))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::CREATED);

    let body: JigPlayerSession = resp.json().await?;

    insta::assert_json_snapshot!(body, { ".**.index" => "[index]", ".**.expires_at" => "[timestamp]" });

    let _resp = client
        .post(&format!("http://0.0.0.0:{}/v1/jig/player", port))
        .json(&serde_json::json!({
            "jigId": "3a71522a-cd77-11eb-8dc1-af3e35f7c743",
            "settings": {
                "direction": "rtl",
                "display_score": true,
                "track_assessments": false,
                "drag_assist": false,
            }
        }))
        .login()
        .send()
        .await?
        .error_for_status()?;

    let resp = client
        .get(&format!(
            "http://0.0.0.0:{}/v1/jig/3a71522a-cd77-11eb-8dc1-af3e35f7c743/player",
            port
        ))
        .login()
        .send()
        .await?;

    assert_eq!(resp.status(), StatusCode::OK);

    let body: JigPlayerSessionListResponse = resp.json().await?;

    insta::assert_json_snapshot!(body, { ".**.index" => "[index]", ".**.expires_at" => "[timestamp]"  });

    Ok(())
}

#[sqlx::test]
async fn session_instance_play_count_flow(
    pool_opts: PgPoolOptions,
    conn_opts: PgConnectOptions,
) -> anyhow::Result<()> {
    let app = initialize_server(
        &[Fixture::MetaKinds, Fixture::User, Fixture::Jig],
        &[],
        pool_opts,
        conn_opts,
    )
    .await;

    let port = app.port();

    tokio::spawn(app.run_until_stopped());

    let client: reqwest::Client = reqwest::ClientBuilder::new()
        .user_agent("mocked user agent")
        .connect_timeout(std::time::Duration::from_secs(5))
        .timeout(std::time::Duration::from_secs(10))
        .build()?;

    let resp = client
        .post(&format!("http://0.0.0.0:{}/v1/jig/player/instance", port))
        .json(&serde_json::json!({
            "index": 1234,
        }))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::CREATED);

    let body: PlayerSessionInstanceResponse = resp.json().await?;

    let token = body.token.clone();

    insta::assert_json_snapshot!(body, {".**.token" => "[instance_token]"});

    let resp = client
        .post(&format!(
            "http://0.0.0.0:{}/v1/jig/player/instance/complete",
            port
        ))
        .json(&serde_json::json!({
            "token": token,
        }))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::NO_CONTENT);

    let resp = client
        .get(&format!(
            "http://0.0.0.0:{}/v1/jig/0cc084bc-7c83-11eb-9f77-e3218dffb008/play-count",
            port,
        ))
        .login()
        .send()
        .await?
        .error_for_status()?;

    let body: serde_json::Value = resp.json().await?;

    insta::assert_json_snapshot!(body);

    Ok(())
}
