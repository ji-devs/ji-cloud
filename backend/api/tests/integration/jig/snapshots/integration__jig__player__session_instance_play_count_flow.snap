---
source: tests/integration/jig/player.rs
expression: body

---
{
  "jigId": "0cc084bc-7c83-11eb-9f77-e3218dffb008",
  "settings": {
    "direction": "ltr",
    "displayScore": true,
    "trackAssessments": true,
    "dragAssist": true
  },
  "token": "[instance_token]"
}
